core/src/main/java/ai/timefold/solver/core/impl/solver/termination/AdaptiveTermination.java

```diff
@@ -1,57 +1,71 @@
+    private double gracePeriodSoftestImprovementDouble;
 
-    private final NavigableMap<Long, Score_> scoresByTime;
+    private final AdaptiveScoreRingBuffer<Score_> scoresByTime;
 
     public AdaptiveTermination(long gracePeriodMillis, double minimumImprovementRatio) {
-        this.gracePeriodMillis = gracePeriodMillis;
+        // convert to nanoseconds here so we don't need to do a
+        // division in the hot loop
+        this.gracePeriodNanos = gracePeriodMillis * NANOS_PER_MILLISECOND;
         this.minimumImprovementRatio = minimumImprovementRatio;
-        this.scoresByTime = new TreeMap<>();
-    }
-
-    private record LevelScoreDiff(boolean harderScoreChanged, double softestScoreDiff) {
-        /**
-         * Calculates the softest score difference between two scores and records
-         * if any harder levels changed. Returns null if the score are the same.
-         *
-         * @param start The first score, typically smaller
-         * @param end The second score, typically larger
-         * @return A {@link LevelScoreDiff} where harderScoreChanged is true
-         *         if and only if a level beside the softest score level changed,
-         *         and the difference between the softest score as a double, or null
-         *         if both scores are the same.
-         * @param <Score_> The score type.
-         */
-        public static <Score_ extends Score<Score_>> @Nullable LevelScoreDiff between(@NonNull Score_ start,
-                @NonNull Score_ end) {
-            var scoreDiffs = end.subtract(start).toLevelDoubles();
-            var softestLevel = scoreDiffs.length - 1;
-            for (int i = 0; i < scoreDiffs.length; i++) {
-                if (scoreDiffs[i] != 0.0) {
-                    return new LevelScoreDiff(softestLevel != i, scoreDiffs[softestLevel]);
-                }
+        this.scoresByTime = new AdaptiveScoreRingBuffer<>();
+    }
+
+    /**
+     * An exception that is thrown to signal a hard score improvement.
+     * This is done since:
+     * <p/>
+     * <ul>
+     * <li/>Hard score improvements are significantly rarer than soft score improvements.
+     * <li/>Allows us to directly return a double in a method instead of using a
+     * carrier type/needing to box it so we can use null as a special value.
+     * <li/>Avoid branching in code on the hot path.
+     * </ul>
+     * <p/>
+     * This exception does not fill the stack trace to improve performance;
+     * it is a signal that should always be caught
+     * and handled appropriately.
+     */
+    private static final class HardLevelImprovedSignal extends Exception {
+        @Override
+        public Throwable fillInStackTrace() {
+            return this;
+        }
+    }
+
+    private static <Score_ extends Score<Score_>> double softImprovement(@NonNull Score_ start,
+            @NonNull Score_ end) throws HardLevelImprovedSignal {
+        if (start.equals(end)) {
```

Comment:
Can the score be _worse_ here? Arguably, it can - sometimes the solver picks a worse move to try to get out of a local optima.

Reply:
No, it uses best score, not step score. Best score should never decrease.
```
stepScope.getPhaseScope().getBestScore()
```